File: src/engines/problemGenerator.ts
import type { Problem, Operation, Difficulty } from './types';

let nextId = 0;

function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function generateId(): string {
  return `problem-${++nextId}-${Date.now()}`;
}

export interface ProblemConfig {
  operation: Operation;
  digitCount?: number;
  difficulty?: Difficulty;
  operandCount?: number; // for addition: 2-5
}

function getDigitCount(config: ProblemConfig): number {
  if (config.digitCount) return config.digitCount;
  switch (config.difficulty ?? 'medium') {
    case 'easy': return 2;
    case 'medium': return 3;
    case 'hard': return 4;
  }
}

export function generateProblem(config: ProblemConfig): Problem {
  const { operation } = config;
  const digitCount = getDigitCount(config);

  switch (operation) {
    case 'addition':
      return generateAdditionProblem(digitCount, config.operandCount);
    case 'subtraction':
      return generateSubtractionProblem(digitCount);
    case 'multiplication':
      return generateMultiplicationProblem(digitCount, config.difficulty);
    case 'shortDivision':
      return generateShortDivisionProblem(digitCount, config.difficulty);
    case 'longDivision':
      return generateLongDivisionProblem(digitCount, config.difficulty);
  }
}

export function createProblemFromOperands(operation: Operation, operands: number[]): Problem {
  let answer: number;
  let remainder: number | undefined;

  switch (operation) {
    case 'addition':
      answer = operands.reduce((s, n) => s + n, 0);
      break;
    case 'subtraction':
      answer = operands[0] - operands[1];
      break;
    case 'multiplication':
      answer = operands[0] * operands[1];
      break;
    case 'shortDivision':
    case 'longDivision':
      answer = Math.floor(operands[1] / operands[0]);
      remainder = operands[1] % operands[0];
      break;
  }

  return {
    id: generateId(),
    operation,
    operands,
    answer: answer!,
    remainder,
  };
}

function generateAdditionProblem(digitCount: number, operandCount?: number): Problem {
  const min = Math.pow(10, digitCount - 1);
  const max = Math.pow(10, digitCount) - 1;
  const count = operandCount ?? 2;

  const operands: number[] = [];
  for (let i = 0; i < count; i++) {
    operands.push(randomInt(min, max));
  }

  return {
    id: generateId(),
    operation: 'addition',
    operands,
    answer: operands.reduce((s, n) => s + n, 0),
  };
}

function generateSubtractionProblem(digitCount: number): Problem {
  const min = Math.pow(10, digitCount - 1);
  const max = Math.pow(10, digitCount) - 1;
  let a = randomInt(min, max);
  let b = randomInt(min, max);
  if (b > a) [a, b] = [b, a];

  return {
    id: generateId(),
    operation: 'subtraction',
    operands: [a, b],
    answer: a - b,
  };
}

function generateMultiplicationProblem(digitCount: number, difficulty?: Difficulty): Problem {
  const min = Math.pow(10, Math.max(1, digitCount - 1));
  const max = Math.pow(10, digitCount) - 1;
  const a = randomInt(min, max);
  const maxMultiplier = difficulty === 'easy' ? 5 : 12;
  const b = randomInt(2, maxMultiplier);

  return {
    id: generateId(),
    operation: 'multiplication',
    operands: [a, b],
    answer: a * b,
  };
}

function generateShortDivisionProblem(digitCount: number, difficulty?: Difficulty): Problem {
  const maxDivisor = difficulty === 'easy' ? 4 : 9;
  const divisor = randomInt(2, maxDivisor);
  const min = Math.pow(10, digitCount - 1);
  const max = Math.pow(10, digitCount) - 1;
  const dividend = randomInt(min, max);

  return {
    id: generateId(),
    operation: 'shortDivision',
    operands: [divisor, dividend],
    answer: Math.floor(dividend / divisor),
    remainder: dividend % divisor,
  };
}

function generateLongDivisionProblem(digitCount: number, difficulty?: Difficulty): Problem {
  const maxDivisor = difficulty === 'easy' ? 15 : difficulty === 'hard' ? 30 : 25;
  const divisor = randomInt(11, maxDivisor);
  const min = Math.pow(10, Math.max(2, digitCount - 1));
  const max = Math.pow(10, digitCount) - 1;
  const dividend = randomInt(min, max);

  return {
    id: generateId(),
    operation: 'longDivision',
    operands: [divisor, dividend],
    answer: Math.floor(dividend / divisor),
    remainder: dividend % divisor,
  };
}

File: go-engines/problem_generator.go
package engines

import (
	"fmt"
	"math/rand"
	"sync/atomic"
	"time"
)

var nextID int64

func generateID() string {
	id := atomic.AddInt64(&nextID, 1)
	return fmt.Sprintf("problem-%d-%d", id, time.Now().UnixMilli())
}

// ProblemConfig configures problem generation.
type ProblemConfig struct {
	Operation    Operation
	DigitCount   int
	NumberSize   NumberSize
	OperandCount int // for addition: 2-5
}

func getDigitCount(config ProblemConfig) int {
	if config.DigitCount > 0 {
		return config.DigitCount
	}
	size := config.NumberSize
	if size == "" {
		size = Medium
	}
	switch size {
	case Small:
		return 2
	case Medium:
		return 3
	case Large:
		return 4
	default:
		return 3
	}
}

func randomInt(min, max int) int {
	return min + rand.Intn(max-min+1)
}

// GenerateProblem generates a random math problem based on the config.
func GenerateProblem(config ProblemConfig) Problem {
	digitCount := getDigitCount(config)

	switch config.Operation {
	case Addition:
		return generateAdditionProblem(digitCount, config.OperandCount)
	case Subtraction:
		return generateSubtractionProblem(digitCount)
	case Multiplication:
		return generateMultiplicationProblem(digitCount, config.NumberSize)
	case ShortDivision:
		return generateShortDivisionProblem(digitCount, config.NumberSize)
	case LongDivision:
		return generateLongDivisionProblem(digitCount, config.NumberSize)
	default:
		return generateAdditionProblem(digitCount, config.OperandCount)
	}
}

// CreateProblemFromOperands creates a Problem from given operands.
func CreateProblemFromOperands(operation Operation, operands []int) Problem {
	var answer int
	var remainder *int

	switch operation {
	case Addition:
		for _, n := range operands {
			answer += n
		}
	case Subtraction:
		answer = operands[0] - operands[1]
	case Multiplication:
		answer = operands[0] * operands[1]
	case ShortDivision, LongDivision:
		answer = operands[1] / operands[0]
		r := operands[1] % operands[0]
		remainder = &r
	}

	return Problem{
		ID:        generateID(),
		Operation: operation,
		Operands:  operands,
		Answer:    answer,
		Remainder: remainder,
	}
}

func generateAdditionProblem(digitCount, operandCount int) Problem {
	min := intPow(10, digitCount-1)
	max := intPow(10, digitCount) - 1
	count := operandCount
	if count < 2 {
		count = 2
	}

	operands := make([]int, count)
	answer := 0
	for i := 0; i < count; i++ {
		operands[i] = randomInt(min, max)
		answer += operands[i]
	}

	return Problem{
		ID:        generateID(),
		Operation: Addition,
		Operands:  operands,
		Answer:    answer,
	}
}

func generateSubtractionProblem(digitCount int) Problem {
	min := intPow(10, digitCount-1)
	max := intPow(10, digitCount) - 1
	a := randomInt(min, max)
	b := randomInt(min, max)
	if b > a {
		a, b = b, a
	}

	return Problem{
		ID:        generateID(),
		Operation: Subtraction,
		Operands:  []int{a, b},
		Answer:    a - b,
	}
}

func generateMultiplicationProblem(digitCount int, numberSize NumberSize) Problem {
	min := intPow(10, maxInt(1, digitCount-1))
	max := intPow(10, digitCount) - 1
	a := randomInt(min, max)

	maxMultiplier := 12
	minMultiplier := 2
	switch numberSize {
	case Small:
		maxMultiplier = 5
	case Large:
		maxMultiplier = 20
		minMultiplier = 5
	}
	b := randomInt(minMultiplier, maxMultiplier)

	return Problem{
		ID:        generateID(),
		Operation: Multiplication,
		Operands:  []int{a, b},
		Answer:    a * b,
	}
}

func generateShortDivisionProblem(digitCount int, numberSize NumberSize) Problem {
	maxDivisor := 9
	switch numberSize {
	case Small:
		maxDivisor = 4
	case Large:
		maxDivisor = 12
	}
	divisor := randomInt(2, maxDivisor)
	min := intPow(10, digitCount-1)
	max := intPow(10, digitCount) - 1
	dividend := randomInt(min, max)

	r := dividend % divisor
	return Problem{
		ID:        generateID(),
		Operation: ShortDivision,
		Operands:  []int{divisor, dividend},
		Answer:    dividend / divisor,
		Remainder: &r,
	}
}

func generateLongDivisionProblem(digitCount int, numberSize NumberSize) Problem {
	maxDivisor := 25
	switch numberSize {
	case Small:
		maxDivisor = 15
	case Large:
		maxDivisor = 30
	}
	divisor := randomInt(11, maxDivisor)
	min := intPow(10, maxInt(2, digitCount-1))
	max := intPow(10, digitCount) - 1
	dividend := randomInt(min, max)

	r := dividend % divisor
	return Problem{
		ID:        generateID(),
		Operation: LongDivision,
		Operands:  []int{divisor, dividend},
		Answer:    dividend / divisor,
		Remainder: &r,
	}
}
